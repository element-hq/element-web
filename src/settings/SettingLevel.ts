/*
Copyright 2024 New Vector Ltd.
Copyright 2020 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

/**
 * Represents the various setting levels supported by the SettingsStore.
 */
export enum SettingLevel {
    // TODO: [TS] Follow naming convention
    DEVICE = "device",
    ROOM_DEVICE = "room-device",
    ROOM_ACCOUNT = "room-account",
    ACCOUNT = "account",
    ROOM = "room",
    PLATFORM = "platform",
    CONFIG = "config",
    DEFAULT = "default",
}
