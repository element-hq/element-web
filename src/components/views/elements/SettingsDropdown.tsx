/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { type JSX, useCallback, useId, useState } from "react";

import SettingsStore from "../../../settings/SettingsStore";
import { type SettingLevel } from "../../../settings/SettingLevel";
import { SETTINGS, type StringSettingKey } from "../../../settings/Settings";
import { useSettingValueAt } from "../../../hooks/useSettings.ts";
import Dropdown, { type DropdownProps } from "./Dropdown.tsx";
import { _t } from "../../../../packages/shared-components/src/utils/i18n.tsx";

interface Props {
    settingKey: StringSettingKey;
    level: SettingLevel;
    roomId?: string; // for per-room settings
    label?: string;
    isExplicit?: boolean;
    hideIfCannotSet?: boolean;
    onChange?(option: string): void;
}

const SettingsDropdown = ({
    settingKey,
    roomId,
    level,
    label: specificLabel,
    isExplicit,
    hideIfCannotSet,
    onChange,
}: Props): JSX.Element => {
    const id = useId();
    const settingValue = useSettingValueAt(level, settingKey, roomId ?? null, isExplicit);
    const [value, setValue] = useState(settingValue);
    const setting = SETTINGS[settingKey];

    const onOptionChange = useCallback(
        (value: string): void => {
            setValue(value); // local echo
            SettingsStore.setValue(settingKey, roomId ?? null, level, value);
            onChange?.(value);
        },
        [settingKey, roomId, level, onChange],
    );

    const disabled = !SettingsStore.canSetValue(settingKey, roomId ?? null, level);
    if (disabled && hideIfCannotSet) return <></>;
    if (!setting.options) {
        console.error("SettingsDropdown used for a setting with no `options`");
        return <></>;
    }

    const label = specificLabel ?? SettingsStore.getDisplayName(settingKey, level)!;
    const options = setting.options.map((option) => {
        return <div key={option.value}>{_t(option.label)}</div>;
    }) as DropdownProps["children"];

    return (
        <div className="mx_SettingsDropdown">
            <label className="mx_SettingsDropdown_label" htmlFor={id}>
                <span className="mx_SettingsDropdown_labelText">{label}</span>
            </label>
            <Dropdown
                id={id}
                onOptionChange={onOptionChange}
                menuWidth={360} // matches CSS width
                value={value}
                disabled={disabled}
                label={label}
            >
                {options}
            </Dropdown>
        </div>
    );
};

export default SettingsDropdown;
