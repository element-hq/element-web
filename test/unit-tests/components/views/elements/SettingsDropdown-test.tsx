/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { render, screen } from "jest-matrix-react";

import { SettingLevel } from "../../../../../src/settings/SettingLevel.ts";
import SettingsDropdown from "../../../../../src/components/views/elements/SettingsDropdown.tsx";

describe("<SettingsDropdown />", () => {
    it("should render a disabled setting", async () => {
        const { asFragment } = render(
            <SettingsDropdown settingKey="Electron.autoLaunch" level={SettingLevel.PLATFORM} />,
        );
        expect(asFragment()).toMatchSnapshot();

        const trigger = screen.getByRole("button");
        expect(trigger).toHaveTextContent("No");
        expect(trigger).toHaveAttribute("aria-disabled", "true");
    });

    it("should not render a disabled setting if hideIfCannotSet=true", async () => {
        const { container } = render(
            <SettingsDropdown settingKey="Electron.autoLaunch" level={SettingLevel.PLATFORM} hideIfCannotSet />,
        );
        expect(container).toBeEmptyDOMElement();
    });

    it("should not render a non-options setting", async () => {
        const { container } = render(<SettingsDropdown settingKey="systemFont" level={SettingLevel.DEVICE} />);
        expect(container).toBeEmptyDOMElement();
    });
});
