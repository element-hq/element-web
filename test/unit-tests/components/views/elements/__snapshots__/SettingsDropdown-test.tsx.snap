// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsDropdown /> should render a disabled setting 1`] = `
<DocumentFragment>
  <div
    class="mx_SettingsDropdown"
  >
    <label
      class="mx_SettingsDropdown_label"
      for="«r0»"
    >
      <span
        class="mx_SettingsDropdown_labelText"
      >
        Open %(brand)s when you log in to your computer
      </span>
    </label>
    <div
      class="mx_Dropdown mx_Dropdown_disabled"
    >
      <div
        aria-describedby="«r0»_value"
        aria-disabled="true"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Open %(brand)s when you log in to your computer"
        aria-owns="«r0»_input"
        class="mx_AccessibleButton mx_Dropdown_input mx_no_textinput mx_AccessibleButton_disabled"
        disabled=""
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Dropdown_option"
          id="«r0»_value"
        >
          <div>
            No
          </div>
        </div>
        <span
          class="mx_Dropdown_arrow"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
